### PR TITLE
Fix issue with unregisterOption

### PR DESCRIPTION
**Pull request by krivaten**
Had an issue where if the model for the options changed, it was firing back an error saying `cannot call method unregisterOption of undefined`. So now I am checking for it prior to calling `unregisterOption`.

### CODE REVIEW

**Comment by Robdel12** [comment]
Can you provide a failing test with this? That would a huge help :)

**Comment by cowboyd** [comment]
@krivaten That is definitely something that should never happen, so if possible, I'd like to see if we can't prevent that condition from ever arising rather than patching it up when it does. Like @Robdel12 said, getting a re-production would really be a huge help.

**Comment by martypenner** [comment]
I've come across this as well. It occurred when I was swapping out the options in the `x-select`, then setting the selected one in the `actions` run loop queue. I'll see if I can narrow down this case and get a jsbin up.

**Comment by martypenner** [comment]
Revisited the project I had this issue on and upgraded ember + ember-cli. I cannot seem to repro this now. Wondering if there was an underlying issue in another lib. @krivaten can you confirm latest ember + ember-cli solves it?

**Comment by JackCA** [comment]
I'm actually running into this as well and am having difficulty creating an integration test to reproduce cleanly.

@Robdel12 @cowboyd would you reconsider merging without test?

**Comment by Robdel12** [comment]
Honestly no, I don't think we'll ever take something without tests. Tests ensure there _is_ an issue (that's reproducible) and it has a confirmed _fix_ for it. Otherwise we could be fixing one thing and creating other bugs. 

That said, maybe we can throw some weight behind this sometime soon to try and see if we can reproduce it. :)

**Comment by JackCA** [comment]
@Robdel12 alright, I'm (still) currently trying to write a failing test for this -- it is most likely related to a timing issue where the wrapping `x-select` is destroyed before the option can be destroyed.

I will update you if I can reproduce in test environment.

**Comment by Robdel12** [comment]
If you'd still like to have this change you can add your own `x-option.js` file in your components dir. Then you can overwrite the `willDestroyElement` hook and add in the check for the time being. 

So create a file like this: `yourapp/app/components/x-option.js` 

Then your `x-option.js` file:

```js
import XOptionComponent from 'emberx-select/components/x-option';

export default XOptionComponent.extend({
  willDestroyElement() {
    this._super.apply(this, arguments);

    var select = this.get('select');
    if (select) {
      select.unregisterOption(this);
    }
  }
});
```

And that should be the temp patch you're looking for :)

**Comment by JackCA** [comment]
@Robdel12 thank you for taking the time to write up that fix -- I really appreciate it.

I've actually got something similar in the code I'm working on -- I'm just trying to contribute back since it looks like at least three other people have had this issue.

I've dug more into this and I can pretty confidently say *why* this is happening, but I still can't get my integration test to fail.

Currently, there is an assumption that the [`registerWithXSelect`](https://github.com/thefrontside/emberx-select/blob/master/addon/components/x-option.js#L59-L64) callback is **always** going to be triggered before the [`willDestroyElement`](https://github.com/thefrontside/emberx-select/blob/master/addon/components/x-option.js#L71-L74) event. However, I don't think that's a safe assumption to make.

It's very possible that the element will be initialized and then immediately destroyed before that `afterRender` event is triggered (meaning there won't be a `select` property). I believe its difficult to recreate with an integration test because you need to have a run loop that will delay the `afterRender` event long enough to demonstrate that this state is definitely possible.

tl;dr: `willDestroyElement` can be called before `afterRender` event has registered the `select` property

**Comment by cowboyd** [comment]
Perhaps an integration test isn't the answer, maybe we can simulate this with a unit test.

**Comment by JackCA** [comment]
@cowboyd that's a good idea but it's difficult to unit test `x-option` because it supposed to be inside of `x-select` and you throw an error in its absence.

We could stub `nearestOfType` but it doesn't look like you have a stubbing library yet so I don't want to make that call.

Even then, I'm not sure how you can enter the state between `inDom` and the `afterRender` state because `this.render()` will get you into `afterRender` immediately (it appears).

**Comment by jcope2013** [comment]
just to chime in, I also came across this and am currently pointing my package.json to this PR

**Comment by cowboyd** [comment]
@JackCA That's the exact case we want to test, right? I'm thinking more like this pseudo-code:

```
it("doesn't blow up if it is destroyed before it has a chance to register with select", function() {
  let option = XComponentOption.create();
  expect(function() { option.willDestroyElement() }).not.to.raise
})
```

**Comment by JackCA** [comment]
@cowboyd implemented here: https://github.com/thefrontside/emberx-select/pull/80

**Comment by Robdel12** [comment]
This was fixed with #80
